lints/lint_mp_ecdsa_allowed_algorithm.go

```diff
@@ -0,0 +1,121 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+/********************************************************************
+Section 5.1 - Algorithms
+ECDSA keys using one of the following curve-hash pairs: P‐256 with SHA-256, P‐384 with SHA-384
+********************************************************************/
+
+package lints
+
+import (
+	"encoding/asn1"
+	"errors"
+	"math/big"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+type ecdsaSignature struct {
+	R, S *big.Int
+}
+
+func getSigningKeySize(cert *x509.Certificate) (int, error) {
+	sig := new(ecdsaSignature)
+	if _, err := asn1.Unmarshal(cert.Signature, sig); err != nil {
+		return -1, err
+	}
+
+	rsize := sig.R.BitLen()
+	ssize := sig.S.BitLen()
+
+	switch {
+	case rsize <= 112 && ssize <= 112:
+		return 112, nil
+	case rsize <= 128 && ssize <= 128:
+		return 128, nil
+	case rsize <= 160 && ssize <= 160:
+		return 160, nil
+	case rsize <= 192 && ssize <= 192:
+		return 192, nil
+	case rsize <= 224 && ssize <= 224:
+		return 224, nil
+	case rsize <= 239 && ssize <= 239:
+		return 239, nil
+	case rsize <= 256 && ssize <= 256:
+		return 256, nil
+	case rsize <= 320 && ssize <= 320:
+		return 320, nil
+	case rsize <= 384 && ssize <= 384:
+		return 384, nil
+	case rsize <= 512 && ssize <= 512:
+		return 512, nil
+	case rsize <= 521 && ssize <= 521:
+		return 521, nil
+	}
+
+	return -1, errors.New("cannot identify signing ECDSA key length")
+}
+
+type ecdsaAllowedAlgorithm struct{}
+
+func (l *ecdsaAllowedAlgorithm) Initialize() error {
+	return nil
+}
+
+func (l *ecdsaAllowedAlgorithm) CheckApplies(c *x509.Certificate) bool {
+	if c.SignatureAlgorithm == x509.ECDSAWithSHA1 || c.SignatureAlgorithm == x509.ECDSAWithSHA256 ||
+		c.SignatureAlgorithm == x509.ECDSAWithSHA384 || c.SignatureAlgorithm == x509.ECDSAWithSHA512 {
+		return true
+	}
+
+	return false
+}
+
+func (l *ecdsaAllowedAlgorithm) Execute(c *x509.Certificate) *LintResult {
+	signKeySize, err := getSigningKeySize(c)
+	if err != nil {
+		return &LintResult{
+			Status:  Fatal,
+			Details: "cannot identify signing ECDSA key length",
+		}
+	}
+
+	switch {
+	case c.SignatureAlgorithm == x509.ECDSAWithSHA256 && signKeySize == 256:
+		return &LintResult{
+			Status:  Pass,
+			Details: "Detected ECDSAWithSHA256 and 256 bit signing key.",
+		}
+	case c.SignatureAlgorithm == x509.ECDSAWithSHA384 && signKeySize == 384:
+		return &LintResult{
+			Status:  Pass,
+			Details: "Detected ECDSAWithSHA384 and 384 bit signing key.",
+		}
+	}
+
+	return &LintResult{Status: Error}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "e_mp_ecdsa_allowed_algorithm",
+		Description:   "ECDSA keys using one of the following curve-hash pairs: P‐256 with SHA-256, P‐384 with SHA-384",
```

Comment:
This name doesn't match the lint logic. The lint is checking the signature, but the description describes the key present in the certificate.

Reply:
The exact test from the policy is:

> ECDSA keys using one of the following curve-hash pairs:
> - P‐256 with SHA-256
> - P‐384 with SHA-384

I used the text as it appears, which I agree is a little bit confusing. When the original description is not exact, is it preferable to follow the text from the requirements or adopt it to actually match the lint?

Reply:
My concern is that the policy is describing how keys in a certificate should be formed whereas the lint is checking signature on the certificate not the public key.

Reply:
How about changing this to "[CA] ECDSA keys [signing certificates] using one of the following curve-hash pairs: P‐256 with SHA-256, P‐384 with SHA-384"? I think that this both describes the actual lint and provides the policy requirement.

Reply:
I don't fully understand the change. That seems to be adding words to the description of the lint to match the logic rather than changing the logic to follow the actual rules. Is there additional context in the Mozilla rules that make this distinction clear that you can provide?

Reply:
Oh, sorry, I didn't understand your original point. The logic of the lint is the actual requirement, however it is not stated well in the policy. There are CAs that have been confused about this too, please check https://bugzilla.mozilla.org/show_bug.cgi?id=1530971 as an example which lead to the following commit that will be integrated in the next mozilla policy version: https://github.com/sleevi/pkipolicy/commit/cc6355ee1fb32500b496d38ffff8abb98e90b3db.

Reply:
If you check the discussion above, the checks were initially in a single lint but per zlint's structure it's best if they are split in different files. Plus, the check you are proposing is not actually making sure that P-256 keys sign certs with ECDSA with SHA256 and P-384 keys sign certs with ECDSA with SHA384. The lint you are proposing is a new one that will apply once 2.7 enters into effect and I believe it should be a new unique lint. The current lint will become more or less a check for the last part of the new 5.1.2.

Reply:
It's indeed a new lint(s), but it is relevant before 2.7 becomes live as the language in the current policy also excludes the use of non RSA or P-256/P-384 for public keys, regardless of what the signing key and algorithm was. Don't feel particularly strongly about one or more check per lint though.

Reply:
>  It's indeed a new lint(s), but it is relevant before 2.7 becomes live as the language in the current policy also excludes the use of non RSA or P-256/P-384 for public keys, regardless of what the signing key and algorithm was

I agree with @tadukurow here that there is an outstanding requirement to flag non-RSA/P-256/P-384 public keys but I don't think we should hold up this PR waiting for a separate lint. The review cycle on this branch has been pretty long and I think we should keep the focus narrow. If folks are amenable I'd suggest we file a separate issue for the new lint and move forward trying to land the lints in this branch now.